source/common/network/address_impl.h

```diff
@@ -53,6 +69,21 @@ class InstanceBase : public Instance {
   const Type type_;
 };
 
+class InstanceFactory {
+public:
+  template <typename InstanceType, typename... Args>
+  static StatusOr<InstanceConstPtr> createInstancePtr(Args&&... args) {
+    absl::Status status;
+    // Use new instead of make_unique here because the instance constructors are private and must be
+    // called directly here.
+    std::unique_ptr<InstanceType> instance(new InstanceType(status, std::forward<Args>(args)...));
```

Comment:
Why are we no longer returning shared ptr but instead unique ptr?

Reply:
Yeah I think I never fully understood the answer to this.

Reply:
Returning a unique_ptr from factory is a good practice because it makes the interface more general. unique_ptr can be implicitly cast to shared_ptr but not vice versa. So return a unique_ptr is equivalent to return a unique_ptr or shared_ptr.

Reply:
I think it's probably better not to change this interface in this PR just because it's better practice. A separate PR can be filed for that. 

Large PRs that change semantics should minimize other changes, especially to interfaces.